src/irc/core/sasl.c

```diff
@@ -28,50 +28,131 @@
 
 #define SASL_TIMEOUT (20 * 1000) // ms
 
-static gboolean sasl_timeout(IRC_SERVER_REC *server)
+static GSList *sasl_mechanisms;
+static GHashTable *sasl_buffers;
+
+void sasl_mechanism_register(const char *name)
 {
-	/* The authentication timed out, we can't do much beside terminating it */
-	irc_send_cmd_now(server, "AUTHENTICATE *");
-	cap_finish_negotiation(server);
+	if (gslist_find_string(sasl_mechanisms, name) == NULL) {
+		sasl_mechanisms = g_slist_append(sasl_mechanisms, g_strdup(name));
+	}
+}
 
-	server->sasl_timeout = 0;
+void sasl_mechanism_unregister(const char *name)
+{
+	GSList *pos;
 
-	signal_emit("server sasl failure", 2, server, "The authentication timed out");
+	pos = gslist_find_string(sasl_mechanisms, name);
+	if (pos != NULL) {
+		g_free(pos->data);
+		sasl_mechanisms = g_slist_remove(sasl_mechanisms, pos->data);
+	}
+}
 
-	return FALSE;
+static gboolean is_registered_mechanism(const char *name)
+{
+	return gslist_find_string(sasl_mechanisms, name) != NULL;
 }
 
-static void sasl_start(IRC_SERVER_REC *server, const char *data, const char *from)
+static void sasl_emit_abort_signal(IRC_SERVER_REC *server)
 {
 	IRC_SERVER_CONNECT_REC *conn;
 
 	conn = server->connrec;
 
-	switch (conn->sasl_mechanism) {
-		case SASL_MECHANISM_PLAIN:
-			irc_send_cmd_now(server, "AUTHENTICATE PLAIN");
-			break;
+	if (is_registered_mechanism(conn->sasl_mechanism)) {
+		char *str;
 
-		case SASL_MECHANISM_EXTERNAL:
-			irc_send_cmd_now(server, "AUTHENTICATE EXTERNAL");
-			break;
+		str = g_strconcat("server sasl abort ", conn->sasl_mechanism, NULL);
```

Comment:
Just make sure `sasl_mechanism` is lowercased when loading it from the config file.